src/sql/explain.rs

```diff
@@ -0,0 +1,482 @@
+                "if {} then {{{}}} else {{{}}}",
+                cond.fmt_with(subqueries),
+                then.fmt_with(subqueries),
+                els.fmt_with(subqueries)
+            ),
+            Exists(..) => {
+                let chain = subqueries.remove(0).nodes.last().unwrap().chain;
+                format!("exists(%{})", chain)
+            }
+            Select(..) => {
+                let chain = subqueries.remove(0).nodes.last().unwrap().chain;
+                format!("select(%{})", chain)
+            }
+        }
+    }
+}
+
+impl AggregateExpr {
+    fn fmt_with(&self, subqueries: &mut Vec<&Explanation>) -> String {
+        format!(
+            "{}({}{})",
+            self.func,
+            if self.distinct { "distinct " } else { "" },
+            self.expr.fmt_with(subqueries)
+        )
+    }
+}
+
+#[derive(Debug)]
+pub struct Separated<'a, T>(&'a str, Vec<T>);
+
+impl<'a, T> std::fmt::Display for Separated<'a, T>
+where
+    T: std::fmt::Display,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
+        for (i, elem) in self.1.iter().enumerate() {
+            if i != 0 {
+                write!(f, "{}", self.0)?;
+            }
+            write!(f, "{}", elem)?;
+        }
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+pub struct Bracketed<'a, T>(&'a str, &'a str, T);
+
+impl<'a, T> std::fmt::Display for Bracketed<'a, T>
+where
+    T: std::fmt::Display,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
+        write!(f, "{}{}{}", self.0, self.2, self.1)
+    }
+}
+
+#[derive(Debug)]
+pub struct Indices<'a>(&'a [usize]);
+
+impl<'a> std::fmt::Display for Indices<'a> {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
+        let mut is_first = true;
+        let mut slice = self.0;
+        while !slice.is_empty() {
+            if !is_first {
+                write!(f, ", ")?;
+            }
+            is_first = false;
+            let lead = &slice[0];
+            if slice.len() > 2 && slice[1] == lead + 1 && slice[2] == lead + 2 {
+                let mut last = 3;
+                while slice.get(last) == Some(&(lead + last)) {
+                    last += 1;
+                }
+                write!(f, "#{}..#{}", lead, lead + last - 1)?;
+                slice = &slice[last..];
+            } else {
+                write!(f, "#{}", slice[0])?;
+                slice = &slice[1..];
+            }
+        }
+        Ok(())
+    }
+}
+
+impl std::fmt::Display for JoinKind {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
+        write!(
+            f,
+            "{}",
+            match self {
+                JoinKind::Inner => "Inner",
+                JoinKind::LeftOuter => "LeftOuter",
+                JoinKind::RightOuter => "RightOuter",
+                JoinKind::FullOuter => "FullOuter",
+            }
+        )
+    }
+}
```

Comment:
It's pretty unfortunate how much of this is basically copied from `expr/explain`. Is it worth trying to split out a utility library? Maybe at least for `Separated` and `Bracketed` and stuff? Seems like maybe even `Explanation` and `ExplanationNode` could be shared with the right generic params.

Reply:
The bulk of the code looks the same but is actually operating on different types, which makes it really hard to be generic. I don't think there would be much compression from trying to share between them unfortunately.

Separated and Bracketed could go in some util library I guess.